### DISCUSSION TITLE
What are the implications of the twinBASIC generics implementation?

### DISCUSSION THREAD

**Discussion by mansellan** [twinbasic/twinbasic#696] [General]
**Background**

[Generics](https://en.wikipedia.org/wiki/Generic_programming) are a way of ensuring type safety. The canonical use-case is for collections, such that a `List (Of Animal)` cannot accept a vegetable as a member. Moreover, `animals.Add(celery)` would be a _compile time_ error, rather than a runtime error.

Generics can help avoid obvious programmer errors, but they can also be much more efficient than treating everything as an `object` because the compiler can reason about the operations that are available.

.Net generics are described as _reified_. I had to Google that word, but it apparently means "treated as a first-class citizen". By contrast, Java generics are "type erased". That means that the runtime knows nothing about generics, but the compiler creates closed types on the fly. This means you can't reason about the generic at runtime.

**Question**

Where does twinBASIC fit on this spectrum? If I understand correctly, the compiler closes generics at the point of use. But at the same time, twinBASIC is a native compiler, so maybe "runtime" is not meaningful.

The Java implementation has widely been regarded as a failure. Can twinBASIC avoid such criticism?

**Comment by bclothier** [comment]
[Wayne's relevant remarks on the question](https://github.com/WaynePhillipsEA/twinbasic/issues/620#issuecomment-1004107271)
> ...I anticipate that the only workable solution to the newly proposed problem might be to introduce generic-constraints, however it is unlikely that constraints will be coming to tB in the near future, as I can foresee it requiring a **lot** of work. At the moment, tB generics are more similar conceptually to C++ templates, as there is no validation done until the generic closed type is encountered and generated....
